File: mobile/test/operators/test_sequence_pool_op.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include <iostream>
#include "../test_include.h"
#include "operators/sequence_ops/sequence_pool_op.h"

namespace paddle_mobile {

int TestSequencePoolOp(const framework::LoDTensor &input_x,
                       const std::string pool_type,
                       framework::LoDTensor *output) {
  VariableNameMap inputs;
  VariableNameMap outputs;
  auto scope = std::make_shared<framework::Scope>();
  inputs["X"] = std::vector<std::string>({"input_x"});
  outputs["Out"] = std::vector<std::string>({"output"});

  auto input_x_var = scope.get()->Var("input_x");
  auto *x = input_x_var->template GetMutable<framework::LoDTensor>();
  x->Resize(input_x.dims());
  x->ShareDataWith(input_x);
  x->set_lod(input_x.lod());

  auto output_var = scope.get()->Var("output");

  framework::AttributeMap attrs;
  attrs["pooltype"].Set<std::string>(pool_type);

  auto *op = new operators::SequencePoolOp<CPU, float>(
      "sequence_pool", inputs, outputs, attrs, scope.get());

  op->InferShape();
  op->Init();
  op->Run();

  auto *out = output_var->template Get<framework::LoDTensor>();
  output->Resize(out->dims());
  output->ShareDataWith(*out);
  delete op;
  return 0;
}

}  // namespace paddle_mobile

// namespace framework = paddle_mobile::framework;

int main(int argc, char *argv[]) {
  framework::LoDTensor input_x, output;
  // case 1
  DLOG << "running max case 1";
  {
    std::vector<float> data{1, 2, 3, 4};
    input_x.Resize(framework::make_ddim({4, 1}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < 4; ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "MAX", &output);
    std::vector<float> expect_data{2, 4};
    for (int i = 0; i < 2; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 2
  DLOG << "running max case 2";
  {
    std::vector<float> data{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    input_x.Resize(framework::make_ddim({data.size(), 1}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 3, 10}});

    TestSequencePoolOp(input_x, "MAX", &output);
    std::vector<float> expect_data{3, 10};
    for (int i = 0; i < 2; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  DLOG << "running max case 3";
  // case 3
  {
    std::vector<float> data{1, 2, 3, 4, 5, 6, 7, 8};
    input_x.Resize(framework::make_ddim({4, 2}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "MAX", &output);
    std::vector<float> expect_data{3, 4, 7, 8};
    for (int i = 0; i < 4; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 4
  DLOG << "running max case 4";
  {
    std::vector<float> data{1,  2,  3,  4,  5,  6,  7,  8,  9,  10,
                            11, 12, 13, 14, 15, 16, 17, 18, 19, 20};
    input_x.Resize(framework::make_ddim({4, 5}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "MAX", &output);
    std::vector<float> expect_data{6, 7, 8, 9, 10, 16, 17, 18, 19, 20};
    for (int i = 0; i < 10; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 1
  DLOG << "running sum case 1";
  {
    std::vector<float> data{1, 2, 3, 4};
    input_x.Resize(framework::make_ddim({4, 1}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < 4; ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "SUM", &output);
    std::vector<float> expect_data{3, 7};
    for (int i = 0; i < 2; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 2
  DLOG << "running sum case 2";
  {
    std::vector<float> data{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    input_x.Resize(framework::make_ddim({data.size(), 1}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 3, 10}});

    TestSequencePoolOp(input_x, "SUM", &output);
    std::vector<float> expect_data{6, 49};
    for (int i = 0; i < 2; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 3
  DLOG << "running sum case 3";
  {
    std::vector<float> data{1, 2, 3, 4, 5, 6, 7, 8};
    input_x.Resize(framework::make_ddim({4, 2}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "SUM", &output);
    std::vector<float> expect_data{4, 6, 12, 14};
    for (int i = 0; i < 4; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 4
  DLOG << "running sum case 4";
  {
    std::vector<float> data{1,  2,  3,  4,  5,  6,  7,  8,  9,  10,
                            11, 12, 13, 14, 15, 16, 17, 18, 19, 20};
    input_x.Resize(framework::make_ddim({4, 5}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "SUM", &output);
    std::vector<float> expect_data{7, 9, 11, 13, 15, 27, 29, 31, 33, 35};
    for (int i = 0; i < 10; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 1
  DLOG << "running first case 1";
  {
    std::vector<float> data{1, 2, 3, 4};
    input_x.Resize(framework::make_ddim({4, 1}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < 4; ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "FIRST", &output);
    std::vector<float> expect_data{1, 3};
    for (int i = 0; i < 2; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 2
  DLOG << "running first case 2";
  {
    std::vector<float> data{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    input_x.Resize(framework::make_ddim({data.size(), 1}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 3, 10}});

    TestSequencePoolOp(input_x, "FIRST", &output);
    std::vector<float> expect_data{1, 4};
    for (int i = 0; i < 2; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 3
  DLOG << "running first case 3";
  {
    std::vector<float> data{1, 2, 3, 4, 5, 6, 7, 8};
    input_x.Resize(framework::make_ddim({4, 2}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "FIRST", &output);
    std::vector<float> expect_data{1, 2, 5, 6};
    for (int i = 0; i < 4; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  // case 4
  DLOG << "running first case 4";
  {
    std::vector<float> data{1,  2,  3,  4,  5,  6,  7,  8,  9,  10,
                            11, 12, 13, 14, 15, 16, 17, 18, 19, 20};
    input_x.Resize(framework::make_ddim({4, 5}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < data.size(); ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});

    TestSequencePoolOp(input_x, "FIRST", &output);
    std::vector<float> expect_data{1, 2, 3, 4, 5, 11, 12, 13, 14, 15};
    for (int i = 0; i < 10; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        DLOG << "output[" << i << "]: " << output.data<float>()[i]
             << " != expect[" << i << "]: " << expect_data[i];
        return 1;
      }
    }
  }
  return 0;
}

File: lite/tests/unittest_py/op/test_sequence_pool_op.py
# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
sys.path.append('../')

from auto_scan_test import AutoScanTest, IgnoreReasons
from program_config import TensorConfig, ProgramConfig, OpConfig, CxxConfig, TargetType, PrecisionType, DataLayoutType, Place
import unittest
from functools import partial
import hypothesis
from hypothesis import given, settings, seed, example, assume
import hypothesis.strategies as st
import numpy as np
import random


class TestSequenceReshapeOp(AutoScanTest):
    def __init__(self, *args, **kwargs):
        AutoScanTest.__init__(self, *args, **kwargs)
        self.enable_testing_on_place(
            TargetType.X86, [PrecisionType.FP32],
            DataLayoutType.NCHW,
            thread=[1, 4])
        self.enable_testing_on_place(
            TargetType.ARM, [PrecisionType.FP32],
            DataLayoutType.NCHW,
            thread=[1, 4])
        self.enable_testing_on_place(
            TargetType.ARM, [PrecisionType.FP16],
            DataLayoutType.NCHW,
            thread=[1, 4])

    def is_program_valid(self,
                         program_config: ProgramConfig,
                         predictor_config: CxxConfig) -> bool:
        return True

    def sample_program_configs(self, draw):
        def generate_lod(seq_num):
            sum = 0
            lod_tensor = []
            lod_tensor.append(sum)
            for i in range(0, seq_num):
                sum += random.randint(1, 9) % 2 + 1
                lod_tensor.append(int(sum))
            return [lod_tensor]

        in_shape = draw(
            st.lists(
                st.integers(
                    min_value=1, max_value=4), min_size=2, max_size=8))
        pad_value = draw(st.sampled_from([0.0, 0.2, 0.5, 1.0]))
        pooltype = draw(
            st.sampled_from(
                ["AVERAGE", "SUM", "SQRT", "LAST", "FIRST", "MAX"]))
        seq_num = draw(st.sampled_from([1, 3, 5]))

        lod_tensor = generate_lod(seq_num)
        in_shape0 = draw(
            st.integers(
                min_value=lod_tensor[0][len(lod_tensor[0]) - 1],
                max_value=lod_tensor[0][len(lod_tensor[0]) - 1] + 8))
        in_shape.insert(0, in_shape0)

        lod_level = draw(st.sampled_from([1, 2]))

        if lod_level == 2:
            in_shape = draw(
                st.lists(
                    st.integers(
                        min_value=1, max_value=6),
                    min_size=2,
                    max_size=4))
            in_shape = [4] + in_shape
            lod_tensor = draw(
                st.sampled_from([[[0, 2, 3], [0, 1, 2, 4]],
                                 [[0, 3, 4], [0, 2, 2, 3, 4]]]))

        def generate_input(*args, **kwargs):
            return np.random.random(in_shape).astype(np.float32)

        ops_config = OpConfig(
            type="sequence_pool",
            inputs={"X": ["input_data"], },
            outputs={"Out": ["output_data"],
                     "MaxIndex": ["maxindex"]},
            attrs={"pad_value": pad_value,
                   "pooltype": pooltype})

        outputs_ = ["output_data"]
        if pooltype == "MAX":
            outputs_ = ["output_data", "maxindex"]

        program_config = ProgramConfig(
            ops=[ops_config],
            weights={},
            inputs={
                "input_data": TensorConfig(
                    data_gen=partial(generate_input), lod=lod_tensor)
            },
            outputs=outputs_)

        return program_config

    def sample_predictor_configs(self):
        atol, rtol = 1e-5, 1e-5
        config_lists = self.get_predictor_configs()
        for config in config_lists:
            if config.precision() in [PrecisionType.FP16]:
                atol, rtol = 1e-3, 1e-3

        return self.get_predictor_configs(), ["sequence_pool"], (atol, rtol)

    def add_ignore_pass_case(self):
        def _teller1(program_config, predictor_config):
            target_type = predictor_config.target()
            pooltype = program_config.ops[0].attrs["pooltype"]
            if target_type == TargetType.ARM and predictor_config.precision(
            ) == PrecisionType.FP16 and pooltype == "MAX":
                return True

        self.add_ignore_check_case(
            _teller1, IgnoreReasons.ACCURACY_ERROR,
            "Lite has diff in a specific case in the 2th output on arm for MAX pooling type. but We needn't fix it."
        )

    def test(self, *args, **kwargs):
        self.run_and_statis(quant=False, max_examples=500)


if __name__ == "__main__":
    unittest.main(argv=[''])
